### PR TITLE
added  contract-to-user-to-ram providers data structure 

### PR DESCRIPTION
**Pull request by e-schepachev**
**Change Description**

**Consensus Changes**

**API Changes**

**Documentation Additions**

### CODE REVIEW

File: libraries/chain/transaction_context.cpp [hunk 1]
```diff
@@ -313,6 +321,13 @@ namespace bacc = boost::accumulators;
       is_initialized = true;
    }
 
+   void transaction_context::add_ram_provider(const action& action) {
```
Comment (s-medvedev) [inline libraries/chain/transaction_context.cpp]: It's unexpected to get action as parameter to this function. I think better pass `provideram`.

File: libraries/chain/include/eosio/chain/transaction_context.hpp [hunk 2]
```diff
@@ -37,6 +37,16 @@ namespace eosio { namespace chain {
...
    };
 
+   struct ram_providers {
+
+      void add_provider(account_name contract, account_name user, account_name provider);
+
+   private:
+       using avalable_providers_for_user = std::map<account_name, std::vector<account_name>>;
+
+       std::map<account_name, avalable_providers_for_user>  available_providers_;
```
Comment (s-medvedev) [inline libraries/chain/include/eosio/chain/transaction_context.hpp]: It's seems that available_providers_ more complex then actually need.
Maybe better `std::map<std::pair<contract,account>, provider>`?
What behavior is planned if for one contract, account there will be several providers? I thin this is an erroneous transaction.

File: libraries/chain/transaction_context.cpp [hunk 3]
```diff
@@ -259,12 +262,14 @@ namespace bacc = boost::accumulators;
              }
          }
       }
-      validate_ram_usage.reserve( bill_to_accounts.size() );
+
```
Comment (s-medvedev) [inline libraries/chain/transaction_context.cpp]: Performance question. But it's better do `validate_ram_usage.reserve()` in this place. `bill_to_accounts` doesn't contains info about account, which will be use provider bandwidth. But this accounts can have ram changes.

File: libraries/chain/transaction_context.cpp [hunk 4]
```diff
@@ -313,6 +318,22 @@ namespace bacc = boost::accumulators;
...
+   }
+
+   void transaction_context::add_ram_provider(account_name contract, account_name user, account_name provider)    {
+       auto provider_for_user = ram_providers_.find(contract);
+
+       if (provider_for_user == ram_providers_.end() || provider_for_user->second.find(user) == provider_for_user->second.end()) {
+           ram_providers_[contract][user] = provider;
+       }
+
+       EOS_THROW(ram_provider_error, "Provider has been already setted up");
```
Comment (s-medvedev) [inline libraries/chain/transaction_context.cpp]: EOS_THROW throws ram_provider_error always.

Comment (e-schepachev) [reply libraries/chain/transaction_context.cpp]: Oh. Throw must be inside if.

File: libraries/chain/include/eosio/chain/contract_types.hpp [hunk 5]
```diff
@@ -174,6 +174,26 @@ struct requestbw {
...
+    provideram(const account_name& provider, const account_name& account, std::vector<name> contracts)
+        : provider(provider), account(account), contracts(std::move(contracts))
+    {}
+
+    static account_name get_account() {
+        return config::system_account_name;
+    }
+
+    static action_name get_name() {
+        return N(providebw);
```
Comment (zxcat) [inline libraries/chain/include/eosio/chain/contract_types.hpp]: maybe `N(provideram)` ?

Comment (e-schepachev) [reply libraries/chain/include/eosio/chain/contract_types.hpp]: yes. fixed in the next commit

**Pull request merged**